### ISSUE TITLE
[Questions] Any reason why `UriData` is final?

### ISSUE THREAD

**Issue by d-markey** [dart-lang/sdk#54333]
Hello,

I just noticed `UriData` has been marked `final` and was wondering why this change?

I've implemented a light weight replacement for data Uris (see https://github.com/d-markey/slim_data_uri / https://pub.dev/packages/slim_data_uri).

The Dart implementation systematically parses/decodes the payload which impacts performance for larger payloads eg. audio (see https://github.com/ryanheise/just_audio/issues/1013 and https://stackoverflow.com/questions/76594319/fix-ui-hang-when-parsing-large-uri-on-flutter-web-without-compute-function/76655361#76655361).

My implementation avoids decoding when the Uri is created, yielding a huge performance boost when creating a data Uri:

```
Uri parsing (base64) - GRAND TOTAL:
  * Total bytes = 108003328
  * Dart Uri = 0:00:00.949091 --> 0.111 KiB/µs
  * SlimDataUri = 0:00:00.000098 --> 1076.245 KiB/µs
  * ratio = 9684.602

Uri parsing (percent) - GRAND TOTAL:
  * Total bytes = 99614720
  * Dart Uri = 0:00:04.640147 --> 0.021 KiB/µs
  * SlimDataUri = 0:00:00.000056 --> 1737.143 KiB/µs
  * ratio = 82859.768
```

Also I'm using Dart 3.2.3 and just retested my code: I was surprised to see it still works even though my class [`SlimUriData`](https://github.com/d-markey/slim_data_uri/blob/main/lib/src/_slim_uri_data.dart) implements `UriData` which is final. Any idea why this works? Is it related to `sdk: ">=2.17.0 <4.0.0"` so maybe `final` doesn't come into play? Does this mean such restrictions can be bypassed by specifying an older SDK?

My package is used by noone, so this change won't do harm :-) But if there's no reason why `UriData` should be final, maybe it could be reverted?

**Comment by eernstg** [comment]
Yes, it works because of [this rule](https://github.com/dart-lang/language/blob/main/accepted/3.0/class-modifiers/feature-specification.md#:~:text=same%20library%2C%20nor-,K%20is%20a%20platform%20library%20and%20L%20is%20a%20pre%2Dfeature%20library.,-//%20a.dart%0Afinal):

> [It is a compile-time error if a direct superinterface is `final` _except_ if ...]
> - _K_ is a platform library and _L_ is a pre-feature library.
>
> [where _K_ declares the `final` class and _L_ declares the subtype]

This means that it will not work any more when you update the language version of your declaration to 3.0 or higher.

@lrhn, WDYT, could there be a way for `SlimUriData` to continue to work?

**Comment by lrhn** [comment]
The reason for making `UriData` final is the same reason used elsewhere: It's a data class, a self-contained object where all state data is publicly visible, and an object with any valid state can be (re-)created using the constructors.

There should not be any need for a subclass, or any other implementation of the same interface.
(And one shouldn't extend it, but rather [favor composition over inheritance](https://en.wikipedia.org/wiki/Composition_over_inheritance).)

That obviously, and maybe erroneously, doesn't consider *performance* as a characteristic.

The idea of keeping some or all of the constructor parameters unparsed and not building a URL string eagerly isn't bad.
Maybe we could do something like that in the Dart `UriData` class. 

Not being validating isn't necessarily an issue, if we can guarantee that all the state we put in, can be encoded into Base-64/percent encoding if necessary. We usually do that, it just costs a little extra in the encoding phase if we need to add more escapes.

But something like `UriData.fromBytes(bytes, lazy: true)` could just store the `bytes` without eagerly encoding them.
If someone then asks for `contentAsBytes`, we can be really quick (probably still need to create a copy, but no parsing needed), and if they ask for a `.uri.toString`, then we can lazily create the URI string when needed.

I think your use-case is also about providing *pre*-Base-64 encoded data, as a string, which means it's not the data that is turned into a URI string, but a substring of that URI slice. In that case not validating can be an issue, because we may detect an error late, which means throwing in a place where users don't expect things to throw.

Or we could make `UriData` non-`final`. Which wouldn't be so weird, since `Uri` isn't `final` either.
I don't think the classes are using non-public APIs of each other, so it's not a safety issue to allow other libraries to implement `UriData`.

**Comment by d-markey** [comment]
> The idea of keeping some or all of the constructor parameters unparsed and not building a URL string eagerly isn't bad. Maybe we could do something like that in the Dart `UriData` class.

If so, feel free to elaborate and take whatever you like from this package. I've only provided it as an example to the [user's problem](https://stackoverflow.com/questions/76594319/fix-ui-hang-when-parsing-large-uri-on-flutter-web-without-compute-function/76655361) but nobody's using it, so it can be quietly disposed of. I don't know how they solved their issue, but if they implemented their own `UriData` then they will have an issue with the `final` modifier too at some point.

> I think your use-case is also about providing _pre_-Base-64 encoded data, as a string, which means it's not the data that is turned into a URI string, but a substring of that URI slice. In that case not validating can be an issue, because we may detect an error late, which means throwing in a place where users don't expect things to throw.

I think the original use-case is to package (audio) assets as strings instead of binary files. I guess it makes the assets readily available to the app.

From my perspective, throwing "just-in-time" when the payload is used as opposed to when the `data:` URI is created, just mimics the way a URI works: if the resource pointed to by the URI is garbage, the program will not throw until the resource has been downloaded only to find the payload is invalid.

Now there's a security concern because not decoding the `data:` URI means it could contain non-base64/percent encoded stuff, eg. malicious JavaScript. So if it later gets integrated into the DOM, and depending on how it is integrated, it could lead to a vulnerability. I've added features in `SlimDataUri` to allow for checking payload safety whithout fully decoding it. It might still be invalid (and cause an exception if decoded), but at least it can be verified that it contains only allowed characters. Other than that, the payload is considered "opaque" until something makes use/sense of it.

> Or we could make `UriData` non-`final`. Which wouldn't be so weird, since `Uri` isn't `final` either. I don't think the classes are using non-public APIs of each other, so it's not a safety issue to allow other libraries to implement `UriData`.

It's your call! I'm not asking for a change, just curious why it was made `final`.
